### PR TITLE
Inputshaping accel plugin

### LINKED ISSUE #374: Provide german translation for input shaping plugin
Do german translation for input shaping plugin.

### PR DESCRIPTION
**Pull request by mfs12**
Closes #374.

### CODE REVIEW

File: src/plugins/Accelerometer/AccelerometerEnums.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+export const AccelStates = {
+  UNKNOWN: 0,
+  ABORTING: 1,
+  INIT: 2,
+  RUNNING: 3,
+  HALTED: 4,
+  ANALYZING: 5
+};
+
+export const AccelAlgorithm = {
```
[the remark this answers is not in the document]
Comment (mfs12) [reply src/plugins/Accelerometer/AccelerometerEnums.js:12]: Yes, it's redundant. Removed it and using now the one from modelEnums.

File: src/plugins/Accelerometer/Accelerometer.vue [hunk 2]
```diff
@@ -101,11 +283,44 @@ import Events from '../../utils/events.js'
...
+import { AccelStates, AccelAlgorithm } from './AccelerometerEnums.js'
+
 
 export default {
 	computed: {
 		...mapState(['selectedMachine']),
-		...mapGetters(['isConnected', 'uiFrozen'])
+		...mapState('machine', ['model']),
+		...mapGetters(['isConnected', 'uiFrozen']),
+		'recorderMenusAxis': function () {
```
[the remark this answers is not in the document]
Comment (mfs12) [reply src/plugins/Accelerometer/Accelerometer.vue:294]: Fixed it for computed and watch functions.

File: src/plugins/Accelerometer/Accelerometer.vue [hunk 3]
```diff
@@ -35,6 +35,136 @@
...
+						<v-col>
+							<v-card-text class="">
+							{{ $t('plugins.accelerometer.filename') }}:<br>
+							<span class="font-weight-bold"> {{ this.recorderFilename }}</span>
+							</v-card-text>
+							<v-card-text label="Movement Command">
+								{{ this.recorder.initCommand }}<br>
+								{{ this.recorder.moveCommand }}<br>
+								{{ this.recorder.testCommand }}<br>
+								movement...
```
[the remark this answers is not in the document]
Comment (mfs12) [reply src/plugins/Accelerometer/Accelerometer.vue:154]: just a obsolete placeholder. removed it.

File: src/plugins/Accelerometer/Accelerometer.vue [hunk 4]
```diff
@@ -119,6 +334,55 @@ export default {
...
+				testCommand: null,
+
+				iteration: 0,
+				algorithm: AccelAlgorithm.NONE,
+
+				tool: null,
+				axis: null,
+				accel: null,
+				param: {
+					// spi specific paramters
```
[the remark this answers is not in the document]
Comment (mfs12) [reply src/plugins/Accelerometer/Accelerometer.vue:361]: As discussed, i completed the list of potential CS and INT pins. Please double check.

**Comment by chrishamm** [comment]
Please restore the original accelerometer i18n keys and add your own inputShaping keys (copied from accelerometer) instead. Alternatively move the i18n stuff to your own i18n JS files and use the new registerPluginLocalization call to register it (just be aware that I haven't tested this function yet).

**Comment by chrishamm** [comment]
OTOH it probably doesn't matter provided the accelerometer plugin will be replaced with it anyway.

First observations:
- Menu icon doesn't seem right, please use the previous one from the accelerometer plugin
- Accelerometer configuration is still present, please remove it
- If a tool isn't named, the combobox shows empty items. See ToolsPanel.vue for reasonable default tool names
- Max Acceleration/Position isn't needed I think? You could display both values as read-only texts somewhere but no need for inputs
- Max Speed isn't used by the generated G1 code. It should generate a proper G1 Fnnn parameter for the first G1 command where Fnnn is the speed in mm/min (not mm/s)
- There is no way to set the number of samples to collect
- The generated code is in the wrong order. M956 should be called after G4 S2 and then the second G1 code should follow
- It is possible to run the sequence even if the selected axis is not homed
- Either make sure the CSV Files are in line with the Accelerometer Analysis panel or even better
  - move the whole content to tab containers -> https://vuetifyjs.com/en/components/tabs/#tab-items
  - add a section for recording profiles
  - add another section for Accelerometer Analysis
- There is no margin between the panels, see https://vuetifyjs.com/en/styles/spacing/

My first attempt didn't produce a usable result so I cannot say if there are any more issues.

File: src/plugins/Accelerometer/Accelerometer.vue (hunk 4 above)
[the remark this answers is not in the document]
Comment (mfs12) [reply src/plugins/Accelerometer/Accelerometer.vue:286]: Removed as advices in a later comment.

**Comment by chrishamm** [comment]
Further notes:
- It is still possible to start profile recording if the selected axis is not homed
- Max Acceleration/Speed/Position do not show any units. Unless they are removed from the final version, please add them
- When FFT analysis is run, look for the highest peak in the amplitudes and prefill input shaping frequency
- There is still no way set the number of samples to collect (M956 S parameter)

I'll send you a patch to wrap the different panels into tabs shortly.

**Comment by mfs12** [comment]
I didn't say i addressed all of the commented points. Sorry, if there was an misunderstanding.

**Comment by mfs12** [comment]
> It is still possible to start profile recording if the selected axis is not homed
I can't start it. It reports an error that axis are not homed.

**Comment by mfs12** [comment]
- [ ] allow profile recording only in printer state "idle"

**Comment by mfs12** [comment]
german translation will not be part of this PR. It will be a separate PR as soon as the alternative internationalization is tested an available.

**Comment by chrishamm** [comment]
I've tested your PR once again and found more issues:

- The "Axis is not homed" notification is barely readable because it times out so quickly. Please use either the default notification timeout (none passed to makeNotification) or use a longer one. In general, error notifications are persistent on DWC so they never time out. Alternatively disable the "Record Profile" button if the selected axis is not homed and display an error text under the Axis combobox (see validation rules above).
- Separate every G-code command with NL. The first G1 command will never run because M-codes can have G parameters, too. This can be avoided if every code gets its own line.
- When the recorded profile is complete, the current tab isn't automatically changed to "Frequency Analysis" even though the profile is selected. Please implement that. Likewise the loaded profile isn't automatically analysed.
- Although my frequency analysis shows a peak 137.6Hz, the prefilled frequency on the IS configuration tab displays 100Hz.

Have you managed to get your accelerometer working with your printer yet? If not, we really need to find a solution next week.

**Comment by mfs12** [comment]
Thanks for the feedback.

It would be very helpful if you comment directly on the code. This way i can see which code your comment relates to. It saves  time and the discussion if needed can be done next to the code.

No, unfortunately it's working yet. I learned a lot about better printing. Mine is only working with very short cables. Either I am totally stupid or it's some hardware problem. Let's chat next week about it.

File: src/store/machine/modelEnums.js [hunk 5]
```diff
@@ -177,6 +177,7 @@ export const MoveShapingType = {
 	ZVD: 'ZVD',
 	ZVDD: 'ZVDD',
 	EI2: 'EI2',
+	EI3: 'EI3',
 	DAA: 'DAA'
```
Comment (chrishamm) [inline src/store/machine/modelEnums.js:181]: Please remove DAA again, it isn't supported any more in v3.4-b2. Also EI3 appears to be missing in the compiled plugin.

Comment (mfs12) [reply src/store/machine/modelEnums.js:181]: should be part of DWC release as it's machine dependent and not a plugin dependency. created another PR to fix this issue. Check

https://github.com/Duet3D/DuetWebControl/pull/380

Comment (chrishamm) [reply src/store/machine/modelEnums.js:181]: Right, thanks for the note. I am afraid that means you have to hard-code that enum into the plugin for now.

Comment (mfs12) [reply src/store/machine/modelEnums.js:181]: fixed in a seperate branch as it's just a temporary fix until next dwc is released.

**Comment by mfs12** [comment]
@dc42 commented:
The logic that decides on the default movement start and end positions seems odd to me. My delta has X axis limits -150 to 150 and it defaults to 60 to 90. My tool changer has limits -185 to 178.5 and it default to 71.4 to 107.1. I was expecting the move to be centred on or to end at the centre of the axis range.

You are sampling during the move instead of after it. That is why you are getting low frequency components. The M956 command should be at the end, after the M400.

- Better to use a long movement rather than a short one. Otherwise, ringing caused by the acceleration at the star of the move may interfere with ringing caused by deceleration, perhaps even cancelling it out at some frequencies.
- Better if the move ends  to too close to the end of the axis, because the ringing may vary along the length of the axis due to changing length of belt segments etc. and printing at the extremes of the axis is less common than printing near the centre.

So best I think to start the move 5mm or 10mm above the axis minimum, and end it e.g. 2/3 of the way between the axis minimum and axis maximum
...taking account of the change to axis minimum/maximum caused by the tool offset, which I think you do already

**Comment by mfs12** [comment]
- [ ] move m400 in before M956
- [ ] improve setting of movement parameters

**Comment by mfs12** [comment]
Closing this PR because the actual work is done at moment at https://github.com/mfs12/DuetWebControl.

If there will be a plan to merge the plugin back a new PR will be setup.

**Pull request closed without merging**